src/core/filters/filters.rs

```diff
@@ -14,10 +14,13 @@ use crate::core::{
     workaround,
 };
 
-use super::meta::filter::FilterMeta;
+use super::{meta::filter::FilterMeta, packets::filter::FilterPacketType};
 
 #[derive(Clone)]
-pub(crate) struct BpfFilter(pub(crate) Vec<u8>);
+pub(crate) struct BpfFilter {
+    pub(crate) key: FilterPacketType,
+    pub(crate) data: Vec<u8>,
```

Comment:
do you mean `FilterPacketType(Vec<u8>)`?
If so, `FilterPacketType` with a value is used both as a key in the hashmap used to retrieve the bytes to rewrite and as a selector for the filtering objects that generate L2 or L3 filters (based on the established criteria, like for example [here](https://github.com/retis-org/retis/pull/229/commits/e6d86fc15e7663d8d3ffca11df35198f146a1381#diff-2c71dfc05d3598b49a9c8b44ce12b66811c5754222ca61b95947fdafe63e4e30R141)).
I factored both things and defined this way to use it both as a selector variant and an actual value (needed for the C-related part).

Reply:
> > do you mean `FilterPacketType(Vec<u8>)`?
> 
> At first I meant:
> 
> ```rust
> enum BpfFilter {
>     L2(Vec<u8>),
>     L3(Vec<u8>),
> }
> ```

I wrote it a bit quick&dirty, but I meant something like that with `FilterPacketType` being `L2` or `L3` variant.

> 
> Or maybe directly in `enum Filter`; but then I though: shouldn't the L2 and the L3 filter be part of the same object, as they make up the `Filter::Packet`. It seems logical to have them both (L3 could be in an Option<>) in one `BpfFilter` object, WDYT?
> 

BpfFilter was initially supposed to be some kind of generic object containing the binary to consume (e.g. one could have been freplaced when we used to dynamic relink filters, and another inlined). Things changed and reusing it for inlining still made sense. Now things are changing again as it happens.

This makes sense (keeps the same intent for `BpfFilter`, assuming it still makes sense) but has the drawback of allowing duplicates while registering (not sure it is bad):
```
pub(crate) enum Filter {
    Packet(L2, BpfFilter), // `get_type()`, `get_data()`
    Packet(L3, BpfFilter), // `get_type()`, `get_data()`
    Meta(FilterMeta),
}
```
```
pub(crate) enum Filter {
    Packet(BpfFilter, Option<BpfFilter>), // `.get_l2()`, `.get_l3()`
    Meta(FilterMeta),
}
```

in this second case we should still keep FilterPacketType::{L2 = magic_l2, L3 = magic_l3} (or they can be simply constants) and we'd probably need some `.get_magic_l{2,3}()`  to perform:
```
if let Filter::Packet(bpf_filter) = filter {
    filters::register_filter(bpf_filter.key as u32, filter)?;
}
```

in `setup_dyn_filters()`

This would not allow duplicates in the `filters` vector (no need).
Some changes are required, maybe more for the second approach.

> (No sure how easy or realistic that would be).

Reply:
> I had to try my proposal to see how it would go, https://paste.ack.tf/e9caa6@raw (untested). I agree, the link between `Filter` and `FilterPacketType` has to be done manually in this version, which isn't 100% perfect.
> 
> I'm starting to think the first proposal would be best too.

of course, map at impl level or match/map where needed. Getter was to give an idea.
I think I'm losing track of first :) 
First of what discussion?

Reply:
> I hope same as the one you were referencing "Slight preference for the first" :)
> 
> ```rust
> pub(crate) enum Filter {
>     Packet(L2, BpfFilter),
>     Packet(L3, BpfFilter),
>     Meta(FilterMeta),
> }
> ```

Yes, that is it.
Just checking as one never knows :)
I think the only non-ideal thing here is the exception in filters, but that's confined to that and after all, reflects the nature of multiple packet filters vs other unique ones.